graphql-kotlin-federation/src/test/kotlin/com/expedia/graphql/federation/FederatedSchemaGeneratorTest.kt

```diff
@@ -0,0 +1,145 @@
+  author: User! @provides(fields : "name")
+  id: String! @external
+  reviews: [Review!]!
+  shippingCost: String! @requires(fields : "weight")
+  weight: Float! @external
+}
+
+type Query {
+  #Union of all types that use the @key directive, including both types native to the schema and extended types
+  _entities(representations: [_Any!]!): [_Entity]!
+  _service: _Service
+}
+
+type Review {
+  body: String!
+  id: String!
+}
+
+type User @extends @key(fields : "userId") {
+  name: String! @external
+  userId: Int! @external
+}
+
+type _Service {
+  sdl: String!
+}
+
+#Federation scalar type used to represent any external entities passed to _entities query.
+scalar _Any
+
+#Federation type representing set of fields
+scalar _FieldSet
+
+#Directs the executor to include this field or fragment only when the `if` argument is true
+directive @include(if: Boolean!) on FIELD | FRAGMENT_SPREAD | INLINE_FRAGMENT
+
+#Directs the executor to skip this field or fragment when the `if`'argument is true.
+directive @skip(if: Boolean!) on FIELD | FRAGMENT_SPREAD | INLINE_FRAGMENT
+
+#Marks the target field/enum value as deprecated
+directive @deprecated(reason: String = "No longer supported") on FIELD_DEFINITION | ENUM_VALUE"""
+
+class FederatedSchemaGeneratorTest {
+
+    @Test
+    fun `verify can generate federated schema`() {
+        val bookResolver = object : FederatedTypeResolver<Book> {
+            override fun resolve(keys: Map<String, Any>): Book {
+                val book = Book(keys["id"].toString())
+                keys["weight"]?.toString()?.toDoubleOrNull()?.let {
+                    book.weight = it
+                }
+                return book
+            }
+        }
+        val userResolver = object : FederatedTypeResolver<User> {
+            override fun resolve(keys: Map<String, Any>): User {
+                val id = keys["userId"].toString().toInt()
+                val name = keys["name"].toString()
+                return User(id, name)
+            }
+        }
+
+        val config = FederatedSchemaGeneratorConfig(
+            supportedPackages = listOf("test.data.queries.federated"),
+            hooks = FederatedSchemaGeneratorHooks(FederatedTypeRegistry(mapOf("Book" to bookResolver, "User" to userResolver)))
+        )
+
+        val schema = toFederatedSchema(config)
+        assertEquals(FEDERATED_SDL, schema.print().trim())
+        val productType = schema.getObjectType("Book")
+        assertNotNull(productType)
+        assertNotNull(productType.getDirective("key"))
+
+        val entityUnion = schema.getType("_Entity") as? GraphQLUnionType
+        assertNotNull(entityUnion)
+        assertTrue(entityUnion.types.contains(productType))
+    }
+
+    @Test
+    fun `verify generator does not add federation queries for non-federated schemas`() {
+        val expectedSchema = """
+            schema {
+              query: Query
+            }
+
+            type Query {
+              _service: _Service
```

Comment:
If the service is not "federate-able" should its SDL contain `_service`?

Reply:
yes -> https://github.com/ExpediaDotCom/graphql-kotlin/pull/280#discussion_r309385781